### PR TITLE
Fix regression with var pattern nullability

### PR DESCRIPTION
**Pull request by jcouv**
There was a regression in `LearnFromDecisionDag` in 16.10 for top-level `var` patterns in https://github.com/dotnet/roslyn/pull/51001

Also fix https://github.com/dotnet/roslyn/issues/52925 (type of `var` should have an annotation) and https://github.com/dotnet/roslyn/issues/46236 (`GetDeclaredSymbol` on such `var` declarations)
Note that any change to the inferred type of `var` is breaking in `ref var` scenario. We've hit this in bootstrap build and I've added a test to illustrate.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs [hunk 1]
```diff
@@ -319,7 +316,7 @@ public PossiblyConditionalState Clone()
             tempMap.Add(rootTemp, (originalInputSlot, expressionType.Type));
 
             var nodeStateMap = PooledDictionary<BoundDecisionDagNode, (PossiblyConditionalState state, bool believedReachable)>.GetInstance();
-            nodeStateMap.Add(decisionDag.RootNode, (state: initialState.Clone(), believedReachable: true));
+            nodeStateMap.Add(decisionDag.RootNode, (state: PossiblyConditionalState.Create(this), believedReachable: true));
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs:319]: :memo: The issue was that `makeDagTempSlot` above was recording the type of the `var` pattern local into `State`, based on correct state (from `expressionType`), but we'd drop that by using `initialState` which captured a prior state.

**Comment by jcouv** [comment]
@RikkiGibson @cston @dotnet/roslyn-compiler for review. Thanks

**Comment by jcouv** [comment]
I'll take a look to see if this also fixed https://github.com/dotnet/roslyn/issues/46236 (Thanks Chuck)

**Review by RikkiGibson — APPROVED** [review]

File: src/Dependencies/Collections/SegmentedDictionary`2.cs [hunk 2]
```diff
@@ -442,7 +442,7 @@ public Enumerator GetEnumerator()
 ConcurrentOperation:
             ThrowHelper.ThrowInvalidOperationException_ConcurrentOperationsNotSupported();
 ReturnFound:
-            ref var value = ref entry._value;
+            ref TValue value = ref entry._value;
```
Comment (RikkiGibson) [inline src/Dependencies/Collections/SegmentedDictionary`2.cs:445]: It wasn't clear to me why this needed to change.

File: src/Compilers/CSharp/Portable/Symbols/TypeWithState.cs [hunk 3]
```diff
@@ -87,7 +87,7 @@ public TypeWithAnnotations ToTypeWithAnnotations(CSharpCompilation compilation,
             if (Type?.IsTypeParameterDisallowingAnnotationInCSharp8() == true)
             {
                 var type = TypeWithAnnotations.Create(Type, NullableAnnotation.NotAnnotated);
-                return State == NullableFlowState.MaybeDefault ? type.SetIsAnnotated(compilation) : type;
+                return (State == NullableFlowState.MaybeDefault || asAnnotatedType) ? type.SetIsAnnotated(compilation) : type;
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Symbols/TypeWithState.cs:90]: Was this all that was needed to fix the 'foreach' scenario?

**Review by cston — APPROVED** [review]

File: src/Dependencies/Collections/SegmentedDictionary`2.cs (hunk 2 above)
Comment (jcouv) [reply src/Dependencies/Collections/SegmentedDictionary`2.cs:445]: Wouldn't bootstrap without this change

File: src/Compilers/CSharp/Portable/Symbols/TypeWithState.cs (hunk 3 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/TypeWithState.cs:90]: Yes, the `for` loop starts with a local declaration. That's handled by `VisitLocalDeclaration`, which correctly calls `ToAnnotatedTypeWithAnnotations` when the local has an inferred type, but when we get to `ToTypeWithAnnotations` the annotation would get dropped.

**Pull request merged**